app/src/main/cpp/skyline/crypto/key_store.cpp

```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: MPL-2.0
+// Copyright © 2020 Skyline Team and Contributors (https://github.com/skyline-emu/)
+
+#include <functional>
+#include <vfs/os_filesystem.h>
+#include "key_store.h"
+
+namespace skyline::crypto {
+    KeyStore::KeyStore(const std::string &rootPath) {
+        vfs::OsFileSystem root(rootPath);
+        if (root.FileExists("title.keys"))
+            ReadPairs(root.OpenFile("title.keys"), &KeyStore::PopulateTitleKeys);
+        if (root.FileExists("prod.keys"))
+            ReadPairs(root.OpenFile("prod.keys"), &KeyStore::PopulateKeys);
+    }
+
+    void KeyStore::ReadPairs(const std::shared_ptr<vfs::Backing> &backing, ReadPairsCallback callback) {
+        std::vector<char> fileContent(backing->size);
+        backing->Read(fileContent.data(), 0, fileContent.size());
+
+        auto lineStart{fileContent.begin()};
+        std::vector<char>::iterator lineEnd;
+        while ((lineEnd = std::find(lineStart, fileContent.end(), '\n')) != fileContent.end()) {
+            auto keyEnd{std::find(lineStart, lineEnd, '=')};
+            if (keyEnd == lineEnd) {
+                throw exception("Invalid key file");
+            }
+
+            std::string_view key(&*lineStart, keyEnd - lineStart);
+            std::string_view value(&*(keyEnd + 1), lineEnd - keyEnd - 1);
+            (this->*callback)(key, value);
+
+            lineStart = lineEnd + 1;
+        }
+    }
+
+    void KeyStore::PopulateTitleKeys(const std::string_view &keyName, const std::string_view &value) {

```

Comment:
String views don't need to be passed by reference either I'm fairly sure, though check godbolt with O3 on that

Reply:
Yeah, there is no difference on godbolt